Include more metadata fields in GraphQL query

mike3985 opened the pull request:
This brings the GraphQL version of rendered news articles pages a lot closer to the regular, Content Store version.

I've explained the changes in detail in the commit message. At a high level they're:

* add more fields to query
* update test fixtures to include the new fields

Question: would this benefit from test coverage? I'm happy to add some. I wrote some in the process of making these changes but a few things about them didn't feel quite right so I wanted to check what the general opinion was before proceeding.

https://trello.com/c/HZRt4r50/1666-make-meta-tags-etc-of-graphql-version-of-news-articles-the-same-as-regular-version, [Jira issue PP-2906](https://gov-uk.atlassian.net/browse/PP-2906)

mike3985 committed:
Include more metadata fields in GraphQL query

This brings the GraphQL version of rendered news articles pages a lot
closer to the regular, Content Store version.

Newly added fields and what they're for:

* `content_id`,
* `public_updated_at`,
* `primary_publishing_organisation`,
* `publishing_app`,
* `rendering_app` and
* `updated_at`
are all used for meta tags of the same name (via the meta_tags component).

* `public_updated_at`
is also used for the "dateModified" attribute of the NewsArticle
structured data (which I think comes from the machine_readable_metadata
component).

* `worldwide_organisations`
  - `base_path` and
  - `title`
enables the use of `worldwide_organisations` in the "From:" byline.

* `organisations`
  - `analytics_identifier`
* `world_locations`
  - `analytics_identifier`
* `worldwide_organisations`
  - `analytics_identifier`
are for populating the "govuk:world-locations" and
"govuk:organisations" meta tags.

* `taxons`
  - `web_url`
is used to populate the "about->sameAs" field in the NewsArticle
structured data.

* `ordered_related_items`,
* `ordered_related_items_overrides` and
* `suggested_ordered_related_items`
used for populating the "Related content" section of the
contextual_sidebar component.

* `withdrawn_notice`
is used for displaying the "This X was withdrawn" notice (using the
notice component).

---

The test fixtures have been updated to reflect the inclusion of these
additional fields. (And a few incidental changes have been made to the
fixtures, too, e.g. minor whitespace changes.)

I'm planning to add more fixtures, and using those to add more test
coverage in later commits.

yndajas commented on app/queries/graphql/edition_query.rb:
```diff
@@ -101,21 +124,35 @@ def query
                   content_id
                   title
                 }
+                worldwide_organisations {
```
**nitpick:** per your previous commit, this could do with a preceding line break

yndajas replied:
Looks like the spaces were kicked along with the other formatting changes, so I'll mark this as resolved

yndajas commented on spec/fixtures/graphql/news_article.json:
```diff
@@ -1,13 +1,18 @@
 {
-  "data":
```
**question:** so are you of the opinion that we don't test within the news article system test spec that all of this is present when rendered using both GraphQL and Content Store? Do we have an alternative, other than relying on our manual diffing checks?

mike3985 replied:
Nope, that's the only alternative I had in mind.

I can add some tests if it makes sense.

mike3985 replied:
I've added GraphQL versions of the News Article model specs. It could be nice to have the same for the system specs, too, but I've left my work-in-progress on that on a separate branch for now https://github.com/alphagov/frontend/compare/graphql-news-articles-meta-tags-system-test?expand=1

KludgeKML replied:
I'm going to reiterate what I said in Slack, which is: I don't think you need request or system tests for GraphQL.

The GraphQL loading is segregated inside the ContentItemLoader, which ultimately resolves (through the ContentItemFactory) into a ContentItem (or subclass) object. Once you get to that point, as long as the model values are identical between ContentStore and GraphQL, anything that's going wrong will go wrong for either source. So if you've proved that the model generated by GraphQL is the same as the one from ContentStore, all the existing system and request tests provide enough coverage already.

yndajas commented on app/queries/graphql/edition_query.rb:
```diff
@@ -12,12 +12,17 @@ def query
           ) {
```
**suggestion:** struggling to parse this from the commit message because of the multiple "for"s and list(s). Is there a way to reword it for clarity?

> `analytics_identifier` for `organisations`, `world_locations` and `worldwide_organisations` for populating the "govuk:world-locations" and "govuk:organisations" meta tags.

mike3985 replied:
Oof, that is a particularly bad one. I'll give it a go 📝

mike3985 commented on app/queries/graphql/edition_query.rb:
```diff
@@ -13,58 +13,74 @@ def query
             ... on Edition {
               base_path
               description
+              document_type
```
Yeah, that theory does make sense. I'll revert this change if there're no votes in favour.

mike3985 committed:
Introduce `fetch_graphql_content_item` test helper

The difference between this helper method and the existing (private)
`fetch_graphql_fixture` method is:

* `fetch_graphql_fixture` returns the equivalent of the response
  payloads we expect from Publishing API's GraphQL endpoint, whereas
* `fetch_graphql_content_item` returns the equivalent of the response
  that we expect from the API adapters method that ContentItemLoader
  uses (`GdsApi.publishing_api.graphql_content_item`)

mike3985 committed:
GraphQL fixtures equivalent to schema examples

We're trying to improve the test coverage of news articles backed by
data from Publishing API's GraphQL endpoint to complement the existing
coverage of news articles backed by data from Content Store.

(We're not confident in this as a long-term solution to the problem, but
the resulting tests are useful enough that we believe it's worth trying
out, living with for a while and forming opinions based on.)

Here we introduce a few new test fixtures representing responses from
Publishing API's GraphQL endpoint. The names of the fixtures and the
choices of content item in the responses themselves are based entirely
on the existing Content Schema examples that we already use in a lot of
Frontend's tests and which themselves represent responses from Content
API.

For that reason, you should find that these files are near-copies of
their respective namesake Content Schema examples, but only including
the fields currently specified in `Graphql::EditionQuery`, because
that's what these responses are intended to reflect.

mike3985 committed:
Make news_article fixtures reflect schema examples

* rename "news_article" test fixture "news_article_with_taxons", to
  reflect its contents and purpose
* rename "news_article_christmas_2016" fixture to just "news_article",
  since its contents are intended to be the Publishing API GraphQL
  version of the existing "news_article" Content Schema example
* update the specs that already used these to use them in a way that's
  closer to the Content Schema examples usage

mike3985 committed:
Move test variables to where they're used

mike3985 committed:
Spec helper for Content Schemas/GraphQL fixtures

And `include` it in all model specs since I'm planning to use the helper
method in some shared examples that are used by model specs.

mike3985 committed:
Fix shared example parameters

This shared example is used in the case study and news article model
specs. In both it's invoked with 2 arguments, but it only had
1 parameter declared on the shared example definition.

I've taken the parameter names from other, similar shared examples.

I don't think in this case it's a big issue, but ignoring the second
argument meant not testing what we thought we were. As can be seen from
the change I've had to make in the news article spec (the specified
example, when it was actually used, didn't pass the test).

mike3985 committed:
Enable use of GraphQL content in shared examples

Since the backend for news articles can either be Content Store or
Publishing API's GraphQL endpoint, we want the tests to reflect that.

This commit changes the shared examples that're used by the news article
model spec. Those shared examples can now either get their source data
from a Content Schemas example (the existing behaviour and still the
default) or alternatively they can use the local spec fixture of the
same name that represents the equivalent response from Publishing API's
GraphQL endpoint.

---

This was a lot of work, which is why I've only done this for the shared
examples that're used directly by the news article model spec so far. If
we like the approach, we can roll it out for more / all of this type of
shared example (and either fixup the changes here or do that work
separately, I'm not fussed).

mike3985 committed:
Run news article model spec for GraphQL as well

This spec, understandably, was making all of its assertions against
a Content Schema example-backed NewsArticle. After this change all of
its tests are run against both the Content Schema example and a local
fixture representing a response from Publishing API's GraphQL endpoint.

yndajas reviewed (dismissed):
This is a lot of work and - in my opinion - a lot of good work!

Re: "a lot", it might be helpful to split this into two or three PRs covering:
- the substantial refactors
- the improved test coverage (if differentiable)
- the introduction of more fields in the query

... but maybe it's not worth the rewrite

leenagupte commented on spec/support/concerns/emphasised_organisations.rb:
```diff
@@ -1,12 +1,13 @@
-RSpec.shared_examples "it can have emphasised organisations" do |document_type, example_name|
+RSpec.shared_examples "it can have emphasised organisations" do |document_type, example_name, data_source: :content_store|
   let(:content_store_response) { GovukSchemas::Example.find(document_type, example_name: example_name) }
+  let(:api_response) { fetch_content_item(document_type, example_name, data_source:) }
```
I'm not sure about the naming here. Is it always an `api_response`? Also it feels a bit generic.

yndajas replied:
I've changed `api_response` to `content_item` throughout

leenagupte commented on spec/system/news_article_spec.rb:
```diff
@@ -126,5 +126,23 @@
+
+    context "when visiting a page in history mode" do
+      let(:path) { "/government/news/final-care-act-guidance-published?graphql=true" }
+      # rubocop:disable RSpec/LetSetup
```
Can you add something to the commit message about why this Cop has been disabled.

KludgeKML replied:
(Particularly, can we double-check this isn't just a before block in disguise)

leenagupte replied:
Yes, agreed. I did wonder why a `let!` is being used in these tests. I've noticed that they've been littered about the tests with no explanation.

leenagupte replied:
Can you make sure that when the tests are added in [graphql-news-articles-meta-tags-etc-remainder](https://github.com/alphagov/frontend/compare/main...graphql-news-articles-meta-tags-etc-remainder) that you add something to the appropriate commit message about why `let!` is being used therefore why `RSpec/LetSetup` is being disabled. I had a quick look and at time of writing that hadn't been done.

yndajas replied:
Yeah I just moved those onto another branch for reference. The intention isn't to open a PR with those changes at the moment given Keith's provocations on our broader approach to testing, but if we do, we'll be sure to address this

yndajas committed:
Improve shared example parameter consistency

This is similar to e7e324698f8865871eaba324681a5063b6d0a1b5, which
updated the emphasised organisations shared examples. In this case
a second argument was being passed to "it can have people" but not "it
has news image". In both of those cases, the used example will remain
the same

Also updated here for consistency is the positional argument (parameter)
naming

Note that the "it can have organisations" example - which is actually in
the emphasised organisations file - is not used anywhere in the codebase

leenagupte commented on spec/models/news_article_spec.rb:
```diff
@@ -1,25 +1,29 @@
+      subject(:news_article) { described_class.new(api_response) }
 
-    let(:content_store_response) { GovukSchemas::Example.find("news_article", example_name: "best-practice-government-response") }
+      let(:api_response) { fetch_content_item("news_article", "best-practice-government-response", data_source: data_source) }
```
A rogue `api_response` has been left in.

leenagupte reviewed (commented):
Thanks for renaming the test variables. One minor inline comment.

KludgeKML reviewed (approved):
LGTM! Thanks for the changes.

KludgeKML commented:
Oh, if you can fix that one variable name before merging (see Leena's comment).
